### ISSUE TITLE
`to_mut` false positive

### ISSUE THREAD

**Issue by clarcharr** [rust-lang-nursery/rust-clippy#1530]
```
warning: methods called `to_*` usually take self by reference; consider choosing a less ambiguous name
  --> src/lib.rs:34:19
   |
34 |     pub fn to_mut(&mut self) -> &mut T {
   |                   ^^^^^^^^^
   |
```

This was found from running Clippy on the [bow](https://docs.rs/bow) crate; note that this follows the same naming convention as [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html).

This lint should require functions starting with `to_mut` to take mutable references instead of immutable.

**Comment by oli-obk** [comment]
weird... why does `Option` take `as_mut` and `Cow` `to_mut`? somewhat inconsistent...

**Comment by mcarton** [comment]
@oli-obk `Option::as_mut` is basically a no-op while `Cow::to_mut` might clone.

**Comment by oli-obk** [comment]
oh right. Yea, this should simply be special-cased on `to_mut`

**Comment by clarcharr** [comment]
Personally I think that `to_mut_X` would be a reasonable convention for a function that would require `&mut self`. Maybe we shouldn't just special case for `to_mut` by itself?

I can't find anything on docs.rs that uses `to_mut_*` though.

**Comment by JoshMcguigan** [comment]
It seems the `to_mut` lint was removed at some point, and based on the playground example below I don't think this false positive still happens. If that all checks out, this issue can be closed.

https://play.rust-lang.org/?gist=13f96cb60d8063ae5394f50bdd39f929&version=nightly&mode=debug&edition=2015

**Comment by flip1995** [comment]
The lint is called `clippy::wrong_self_convention`, not `to_mut`. The Playground example still produces this warning.

**Comment by JoshMcguigan** [comment]
I'm not seeing that lint being triggered in the playground link I posted above. Is there another playground link which demonstrates this false positive?

**Comment by flip1995** [comment]
![image](https://user-images.githubusercontent.com/9744647/46258086-197ca280-c4c5-11e8-8b81-1b4b7be951a2.png)
Do you run Clippy with Tools->Clippy?

Here is a [Playground](https://play.rust-lang.org/?gist=2a16a64958c5ef3d5ce6c55263dd9951&version=nightly&mode=debug&edition=2015) link with every other warning suppressed.

**Comment by JoshMcguigan** [comment]
Ah, that explains it. Thanks for pointing that out.

**Comment by JoshMcguigan** [comment]
I've started working on this, but I'm not sure what behavior we want. 

It is clear to me that `fn to_mut(&mut self) {}` should NOT trigger the lint. But do we want the lint to be triggered by other usages of `to_mut`? For example, should `fn to_mut(&self) {}` trigger the lint? Also, do we want to extend these rules to cover `to_mut_*`, or should they only cover `to_mut`?
